pkg/controller/encryption/ipsec/ipsec_controller.go

```diff
@@ -0,0 +1,526 @@
+	}
+	return nil
+}
+
+func (c *IPSecController) detachTcDecrypt() error {
+	nodeNsPath := kmesh_netns.GetNodeNSpath()
+	detachFunc := func(netns.NetNS) error {
+		return c.handleTc(constants.TC_DETACH)
+	}
+
+	if err := netns.WithNetNSPath(nodeNsPath, detachFunc); err != nil {
+		log.Errorf("failed to exec tc program detach, %v", err)
+	}
+	return nil
+}
+
+func (c *IPSecController) handleKNIAdd(obj interface{}) {
+	kni, ok := obj.(*v1alpha1.KmeshNodeInfo)
+	if !ok {
+		log.Errorf("expected *v1alpha1_core.KmeshNodeInfo but got %T in handle add func", obj)
+		return
+	}
+
+	if kni.Name == c.kmeshNodeInfo.Name {
+		return
+	}
+	c.queue.AddRateLimited(kni.Name)
+}
+
+func (c *IPSecController) handleKNIUpdate(oldObj, newObj interface{}) {
+	newKni, okNew := newObj.(*v1alpha1.KmeshNodeInfo)
+	if !okNew {
+		log.Errorf("expected *v1alpha1_core.KmeshNodeInfo but got %T in handle update new obj func", newObj)
+		return
+	}
+
+	oldKni, okold := oldObj.(*v1alpha1.KmeshNodeInfo)
+	if !okold {
+		log.Errorf("expected *v1alpha1_core.KmeshNodeInfo but got %T in handle update old obj func", oldObj)
+		return
+	}
+
+	if newKni.Name == c.kmeshNodeInfo.Name {
+		return
+	}
+
+	if reflect.DeepEqual(oldKni.Spec, newKni.Spec) {
+		return
+	}
+
+	c.queue.AddRateLimited(newKni.Name)
+}
+
+func (c *IPSecController) handleKNIDelete(obj interface{}) {
+	node, ok := obj.(*v1alpha1.KmeshNodeInfo)
+	if !ok {
+		log.Errorf("expected *v1alpha1_core.KmeshNodeInfo but got %T in handle delete func", obj)
+		return
+	}
+	nodeNsPath := kmesh_netns.GetNodeNSpath()
+	deleteFunc := func(netns.NetNS) error {
+		for _, targetIP := range node.Spec.Addresses {
+			c.ipsecHandler.mutex.Lock()
+			c.ipsecHandler.Clean(targetIP)
+			c.ipsecHandler.mutex.Unlock()
+		}
+		return nil
+	}
+	netns.WithNetNSPath(nodeNsPath, deleteFunc)
+	for _, addr := range node.Spec.Addresses {
+		for _, cidr := range c.ipsecHandler.nodeInfos[addr].cidrs {
+			c.deleteKNIMapCIDR(cidr, c.kniMap)
+		}
+	}
+}
+
+func (c *IPSecController) handleOneNodeInfo(node *v1alpha1.KmeshNodeInfo) error {
+	// can't change ipsec informationo when process
+	c.ipsecHandler.mutex.Lock()
+	defer c.ipsecHandler.mutex.Unlock()
+
+	nodeNsPath := kmesh_netns.GetNodeNSpath()
+
+	handleFunc := func(netns.NetNS) error {
+		for _, remoteNicIP := range node.Spec.Addresses {
+			c.ipsecHandler.SetNodeInfo(remoteNicIP, node.Spec.BootID, node.Spec.SPI, false, node.Spec.PodCIDRs)
+			for _, localNicIP := range c.kmeshNodeInfo.Spec.Addresses {
+				if err := c.ipsecHandler.CreateXfrmRuleIngress(remoteNicIP, localNicIP); err != nil {
+					return err
+				}
+				if err := c.ipsecHandler.CreateXfrmRuleEgress(localNicIP, remoteNicIP); err != nil {
+					return fmt.Errorf("create xfrm out rule failed, %v", err)
+				}
+			}
+		}
+		return nil
+	}
+	if err := netns.WithNetNSPath(nodeNsPath, handleFunc); err != nil {
+		return err
+	}
+
+	for _, addr := range node.Spec.Addresses {
+		for _, cidr := range c.ipsecHandler.nodeInfos[addr].cidrs {
```

Comment:
Why do not make node name as the key?

Reply:
You can also use the name of the node as the key. In the current processing process, such as creating state rules, the iphandler processes information based on IP addresses and uses IP addresses to classify information.